tensorrt_llm/_torch/auto_deploy/transformations/library/rope.py

```diff
@@ -0,0 +1,383 @@
+        if need_transpose:
+            q_for_op = graph.call_function(torch.ops.aten.transpose, args=(q_node, 1, 2))
+            k_for_op = graph.call_function(torch.ops.aten.transpose, args=(k_node, 1, 2))
+            q_for_op_contig = graph.call_method("contiguous", (q_for_op,))
+            k_for_op_contig = graph.call_method("contiguous", (k_for_op,))
+        else:
+            q_for_op_contig, k_for_op_contig = q_node, k_node
+
+        flash_node = graph.call_function(
+            torch.ops.rope.flashinfer,
+            args=(q_for_op_contig, k_for_op_contig, cos_node, sin_node, True),
+        )
+
+    with graph.inserting_after(flash_node):
+        raw_q = graph.call_function(operator.getitem, args=(flash_node, 0))
+        raw_k = graph.call_function(operator.getitem, args=(flash_node, 1))
+
+    if need_transpose:
+        with graph.inserting_after(raw_q):
+            new_q = graph.call_function(torch.ops.aten.transpose, args=(raw_q, 1, 2))
+        with graph.inserting_after(raw_k):
+            new_k = graph.call_function(torch.ops.aten.transpose, args=(raw_k, 1, 2))
+    else:
+        new_q, new_k = raw_q, raw_k
+
+    q_match["add_node"].replace_all_uses_with(new_q)
+    k_match["add_node"].replace_all_uses_with(new_k)
+
+
+def _process_rope_v2(graph, q_match, k_match, start_boundary):
+    """
+    Process a region that matched the complex-multiplication RoPE pattern (v2).
+    Inserts the custom op (flashinfer) after extracting frequency info and replaces
+    the original type_as nodes.
+    """
+    q_node = q_match["input"]
+    k_node = k_match["input"]
+    inv_freq_node = q_match["inv_freq"]
+
+    if inv_freq_node != k_match["inv_freq"]:
+        raise RuntimeError("Mismatch of freqs_cis (inv_freq) between branches.")
+
+    # Sanity check that input layout is BSND (no transpose needed).
+    q_fake = q_node.meta.get("val", None)
+    if q_fake is not None and len(q_fake.shape) > 2:
+        if not (isinstance(q_fake.shape[1], torch.SymInt) and isinstance(q_fake.shape[2], int)):
+            ad_logger.warning(
+                f"""Sanity check failed: q_fake should have shape [b, s, n, d],
+                s should be symbolic and n should be int, instead got shape {q_fake.shape}"""
+            )
+    else:
+        ad_logger.warning(
+            f"Sanity check failed: q_fake should be 3D or 4D, but got shape {q_fake.shape}"
+        )
+
+    with graph.inserting_before(q_match["out"]):
+        q_for_op_contig, k_for_op_contig = q_node, k_node
+        cos_from_freqs = graph.call_function(torch.ops.aten.real, args=(inv_freq_node,))
+        sin_from_freqs = graph.call_function(torch.ops.aten.imag, args=(inv_freq_node,))
+        cos_flash = graph.call_function(
+            torch.ops.aten.cat, args=((cos_from_freqs, cos_from_freqs), -1)
+        )
+        sin_flash = graph.call_function(
+            torch.ops.aten.cat, args=((sin_from_freqs, sin_from_freqs), -1)
+        )
```

Comment:
There is a chance we can "reuse" the pre-processing of `cos` and `sin` and only do it once per graph. Specifically we can do the following. 

You can maintain a lookup table of nodes from `inv_freq_node -> (cos_flash, sin_flash)`. Every time you start from the same `inv_freq_node` you just look up the corresponding `(cos_flash, sin_flash)` instead of recomputing it. If the node is not already in the lookup you add the nodes into the graph and store it in the lookup table.

Note that you can also combine this with my other comment I made above. 

So probably you want a lookup that is something like `inv_freq_node -> cos_sin_flash` directly.

Reply:
Also note that you will only need to add one `position_ids` ever and can reuse it for every call to the rope op

Reply:
Note that you can apply a similar strategy for v1 rope